Step3 리뷰 다시 요청드립니다

m3252 opened the pull request:
이미지가 저작권이 있을줄 몰랐네요. 감사합니다!

m3252 committed:
docs: README.md 요구사항 정의

m3252 committed:
refactor: delete 기능 수정 및 엔티티 매핑 수정

m3252 committed:
refactor: 익셉션 재정의

m3252 committed:
refactor: qna 삭제하기 리팩토링

m3252 committed:
refactor: 생성자에 생성일자 제거

m3252 committed:
test: 질문 삭제하기 리팩토링 후 테스트

wheejuni commented on README.md:
```diff
@@ -0,0 +1,73 @@
+
+
+# JPA
```
네, 저작권에 관한 부분은 이미지에 대해서만 말씀드린 건 아닙니다. 강의자료 전반에 관한 말씀이었구요,
지금처럼 강의자료의 일부 혹은 전체를 복사 붙여넣기하여 공개 저장소에 탑재하시는 방법은 피해주시기를 부탁드립니다. 
강의자료를 꼭 레포에 담고 싶으시다면 요약이나 가공을 거친 후 올려주시면 좋겠고요, 지금 올려주신 부분도 사실은 너무 과합니다. 
앞으로 수료시까지 유념해주시고 이 부분은 조금만 신경써주시기를 부탁드립니다.

wheejuni commented on src/main/java/qna/CannotDeleteException.java:
```diff
@@ -1,6 +1,6 @@
 package qna;
 
-public class CannotDeleteException extends Exception {
+public class CannotDeleteException extends RuntimeException {
```
👍

wheejuni commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,27 @@
+@Embeddable
+public class Answers {
+
+    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY)
```
`@Embeddable` 안에서 `fetch` 를 `LAZY` 로 두신 이유가 있을까요?
어떤 측면에서 이런 매핑을 구상하셨는지 궁금합니다.

m3252 replied:
생각없이 예전 JPA를 배웠을때 Lazy 로 일단 설정해두라는 내용이 있었던것 같아서 습관적으로 그렇게 매핑을 시킨것 같습니다.

answer는 항상 answers에 포함되기에 answer만 가져올일은 없었는데.. 따로 구상했던 내용은 아닙니다 ㅠ

wheejuni commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,27 @@
+        answers.add(answer);
+    }
+
+    public void forEach(Consumer<? super Answer> action) {
+        Objects.requireNonNull(action);
+        for (Answer answer : answers) {
+            action.accept(answer);
+        }
+    }
```
단순히 `action` 을 바로 받는것은 이 컬렉션 객체의 책임을 다 하지 않는 설계로 생각됩니다.
어떤 액션이 필요한지 구상해보시고, 해당 액션에 대한 책임을 온전히 `Answers` 객체가 가져야 하는 건 아닌지 생각되네요.

wheejuni commented on src/main/java/qna/domain/Question.java:
```diff
@@ -62,18 +71,20 @@ public String getContents() {
 
-    public void delete(boolean deleted) {
-        this.deleted = deleted;
+    public void delete(User writer, List<DeleteHistory> deleteHistories) {
```
파라메터로 `deleteHistories` 를 받기 보다는 이 메소드 안에서 새로운 리스트 객체가 형성되었으면 좋겠습니다. 
파라메터로 넘어온 컬렉션 객체에 원소를 더하고 빼는 일은 대부분의 경우 그렇게 권장되지 않습니다. 불필요한 사이드 이펙트가 발생할 우려가 있고, 이렇게 되면 디버깅하기도 굉장히 까다로워질 겁니다.

wheejuni commented on src/main/java/qna/domain/Question.java:
```diff
@@ -62,18 +71,20 @@ public String getContents() {
+    public void delete(User writer, List<DeleteHistory> deleteHistories) {
+        this.deleted = validateOwner(writer);
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, getId(), getWriter()));
+        answers.forEach(answer -> {
+            answer.delete(writer);
+            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter()));
+        });
```
stream API가 제공하는 아래 메소드들을 사용해 리팩토링 시도해보시면 좋을 것 같아요.

- `.peek()`
- `.map()`
- `.collect()`

wheejuni reviewed (changes requested):
대단히 수고 많으셨습니다. 👍 
몇가지 피드백을 남겨드렸습니다. 확인 및 수정을 부탁드릴게요.

m3252 committed:
feat: 삭제 히스토리 일급객체 생성

m3252 committed:
refactor: 피드백 반영
- forEach 제거하고 delete() 메서드를 Answers 가 갖도록 수정
- stream API 를 통해 리팩토링 시도해보기
- 인자로 컬렉션을 받아와 변경하지 않도록 수정

m3252 committed:
refactor : 메소드 행위 분리

wheejuni reviewed (approved):
수고 많으셨습니다. 👍
